packages/valory/skills/abstract_round_abci/utils.py

```diff
@@ -147,3 +167,302 @@ def parse_tendermint_p2p_url(url: str) -> Tuple[str, int]:
         port = DEFAULT_TENDERMINT_P2P_PORT
 
     return hostname, port
+
+
+##
+# Typing utils - to be extracted to open-aea
+##
+
+
+try:
+    # Python >=3.8 should have these functions already
+    from typing import get_args as _get_args  # pylint: disable=ungrouped-imports
+    from typing import get_origin as _get_origin  # pylint: disable=ungrouped-imports
+except ImportError:
+    # Python 3.7
+    def _get_origin(tp):  # type: ignore
+        """Copied from the Python 3.8 typing module"""
+        if isinstance(tp, typing._GenericAlias):  # pylint: disable=protected-access
+            return tp.__origin__
+        if tp is typing.Generic:
+            return typing.Generic
+        return None
+
+    def _get_args(tp):  # type: ignore
+        """Copied from the Python 3.8 typing module"""
+        if isinstance(tp, typing._GenericAlias):  # pylint: disable=protected-access
+            res = tp.__args__
+            if get_origin(tp) is collections.abc.Callable and res[0] is not Ellipsis:
+                res = (list(res[:-1]), res[-1])
+            return res
+        return ()
+
+
+def get_origin(tp):  # type: ignore
+    """
+    Get the unsubscripted version of a type.
+
+    This supports generic types, Callable, Tuple, Union, Literal, Final and
+    ClassVar. Returns None for unsupported types.
+    Examples:
+        get_origin(Literal[42]) is Literal
+        get_origin(int) is None
+        get_origin(ClassVar[int]) is ClassVar
+        get_origin(Generic) is Generic
+        get_origin(Generic[T]) is Generic
+        get_origin(Union[T, int]) is Union
+        get_origin(List[Tuple[T, T]][int]) == list
+    """
+    return _get_origin(tp)
+
+
+def get_args(tp):  # type: ignore
+    """
+    Get type arguments with all substitutions performed.
+
+    For unions, basic simplifications used by Union constructor are performed.
+    Examples:
+        get_args(Dict[str, int]) == (str, int)
+        get_args(int) == ()
+        get_args(Union[int, Union[T, int], str][int]) == (int, str)
+        get_args(Union[int, Tuple[T, int]][str]) == (int, Tuple[str, int])
+        get_args(Callable[[], T][int]) == ([], int)
+    """
+    return _get_args(tp)
+
+
+##
+# The following is borrowed from https://github.com/tamuhey/dataclass_utils/blob/81580d2c0c285081db06be02b4ecdd125532bef5/dataclass_utils/type_checker.py#L152
```

Comment:
The below is actually better than a lot of other checks I found that are shorter in loc.

It handles well nested types like `Optional[List[int]]` and many other complex types.

Reply:
I extended the implement and fixed a bunch of issues in it

Reply:
sounds rough. What issues did you "fix" precisely? Sounds like you need to add tests to ensure this monkey-patched code remains functional; whether the original implementation changes.

Reply:
There are tests here: https://github.com/valory-xyz/open-autonomy/pull/1731/files#diff-53e2b3e55950193314b7944de0a275bd4d0b219930704ff338698b2a68881582R595-R650 . You can propose more. I'd also like a cleaner solution but after searching for a few days and testing various ones this is the best one I could find.

Reply:
I added the handling of type hints like `Type[SomeClass]`.